test/services/postcode_service_test.rb

```diff
@@ -0,0 +1,37 @@
+require "test_helper"
+
+describe PostcodeService do
+  describe "#valid?" do
+    it "returns true if postcode is valid" do
+      postcode = "E1 8QS"
+      assert(described_class.new(postcode).valid?)
+    end
+
+    it "returns false if the postcode is not valid" do
+      postcode = "invalid postcode"
+      assert_not(described_class.new(postcode).valid?)
+    end
+  end
+
+  describe "#sanitize" do
+    it "strips trailing spaces from entered postcodes" do
+      assert_equal "WC2B 6NH", described_class.new("WC2B 6NH ").sanitize
+    end
+
+    it "transposes O/0 and I/1 if necessary" do
+      assert_equal "W1A 0AA", described_class.new("WIA OAA").sanitize
+    end
+
+    it "returns nil if the postcode parameter is nil or an empty string" do
+      assert_nil described_class.new("").sanitize
+    end
+
+    it "removes special characters from the start of the postcode" do
+      assert_equal "WC2B 6NH", described_class.new(".WC2B6NH").sanitize
```

Comment:
I wonder if it's worth testing specifically for the non-breaking space that causes https://sentry.io/organizations/govuk/issues/1954296277/?project=202213&query=is%3Aunresolved? 

(Calls `<MAPIT>/postcode/%C2%A0WC1A1AAjson`)

Reply:
Interestingly, "%C2%A0WC1A1AA" is sanitized as "C2A0WC1A1AA" 😅 , though I don't think that's quite the same as the value the user is typing in.

Reply:
Sorry 😬

Reply:
Looks great otherwise :-D

Reply:
That's ok, I found the right character! `\u00A0` I'll add a test for it.